packages/techdocs-common/src/stages/publish/googleStorage.ts

```diff
@@ -0,0 +1,222 @@
+      });
+
+    return new GoogleGCSPublish(storageClient, bucketName, logger);
+  }
+
+  constructor(
+    private readonly storageClient: Storage,
+    private readonly bucketName: string,
+    private readonly logger: Logger,
+  ) {
+    this.storageClient = storageClient;
+    this.bucketName = bucketName;
+    this.logger = logger;
+  }
+
+  /**
+   * Upload all the files from the generated `directory` to the GCS bucket.
+   * Directory structure used in the bucket is - entityNamespace/entityKind/entityName/index.html
+   */
+  publish({ entity, directory }: PublishRequest): Promise<{}> {
+    return new Promise((resolve, reject) => {
+      // Path of all files to upload, relative to the root of the source directory
+      // e.g. ['index.html', 'sub-page/index.html', 'assets/images/favicon.png']
+      const allFilesToUpload: Array<string> = [];
+
+      // Iterate on all the files in the directory and its sub-directories
+      walk(directory)
+        .on('data', (item: walk.Item) => {
+          // GCS manages creation of parent directories if they do not exist.
+          // So collecting path of only the files is good enough.
+          if (item.stats.isFile()) {
+            // Remove the absolute path prefix of the source directory
+            const relativeFilePath = item.path.replace(`${directory}/`, '');
+            allFilesToUpload.push(relativeFilePath);
+          }
+        })
+        .on('error', (err: Error, item: walk.Item) => {
+          const errorMessage = `Unable to read file at ${item.path}. Error ${err.message}`;
+          this.logger.error(errorMessage);
+          reject(errorMessage);
+        })
+        .on('end', () => {
+          // 'end' event happens when all the files have been read.
+          const entityRootDir = `${entity.metadata.namespace}/${entity.kind}/${entity.metadata.name}`;
+          allFilesToUpload.forEach(filePath => {
+            const source = path.join(directory, filePath); // Local file absolute path
+            const destination = `${entityRootDir}/${filePath}`; // GCS Bucket file relative path
+            this.storageClient
+              .bucket(this.bucketName)
+              .upload(source, { destination })
+              .then(
+                (uploadResp: UploadResponse) => ({
+                  fileName: destination,
+                  status: uploadResp[0],
+                }),
+                (err: Error) => {
+                  const errorMessage = `Unable to upload file ${destination} to GCS. Error ${err.message}`;
+                  this.logger.error(errorMessage);
+                  reject(errorMessage);
+                },
+              );
+          });
+
+          this.logger.info(
+            `Successfully uploaded all the generated files for Entity ${entityRootDir}. Total number of files: ${allFilesToUpload.length}`,
+          );
+          resolve({});
+        });
+    });
+  }
+
+  fetchTechDocsMetadata(entityName: EntityName): Promise<string> {
```

Comment:
Would it be useful to return a typed object here instead of a string?

Reply:
Yes, agreed! Currently there is no type definition for TechDocs Metadata and there should be one. I'll follow up by creating an issue about it.